### ISSUE TITLE
dash::Shared broken for composite types

### ISSUE THREAD

**Issue by fuerlinger** [dash-project/dash#336]
The following does not compile any more because `dash::Shared` insists on keeping a  `GlobRef<Atomic<ElementType>>`  as a member.

~~~
struct test
{
  int a;
  double b;
};

int main(int argc, char* argv[])
{
  dash::init(&argc, &argv);

  dash::Shared<test> s;
  
  dash::finalize();
}
~~~

**Comment by fuchsto** [comment]
@fmoessbauer Type traits for `dash::Atomic` are missing. Then, `dash::Shared<T>::atomic_ref_type` should be:

```c++
using atomic_ref_type = typename std::conditional<
                           dash::is_atomic_value<T>::value,
                           GlobRef<Atomic<T>>,
                           void
                        >::type

```

**Comment by fuchsto** [comment]
... or, fundamentally: `dash::Shared` should not have an `atomic` accessor. A shared global either is atomic, or it is not. Otherwise the owner unit could perform non-atomic changes to the shared value.

For atomic operations on a `dash::Shared`, use `dash::Shared<dash::Atomic<T>>`.

**Comment by fmoessbauer** [comment]
I agree on that, but IMO this should not be implemented or forced using a type trait. Please have a look at my comment in [PR 338](https://github.com/dash-project/dash/pull/338#discussion_r107706607)

**Comment by fuchsto** [comment]
In the current version of the PR, I just removed `dash::Shared.atomic`.

**Comment by fuchsto** [comment]
Fixed in PR #338, also added test of `dash::Shared` for composite value types.
